polynote-frontend/polynote/cell.js

```diff
@@ -407,6 +423,52 @@ export class CodeCell extends Cell {
     get content() {
         return this.editor.getValue();
     }
+
+    applyEdits(edits) {
+        // can't listen to these edits or they'll be sent to the server again
+        // TODO: is there a better way to silently apply these edits? This seems like a hack; only works because of
+        //       single-threaded JS, which I don't know whether workers impact that assumption (JS)
+        this.applyingServerEdits = true;
+
+        try {
+            const monacoEdits = [];
+            const model = this.editor.getModel();
+            edits.forEach((edit) => {
+                if (edit.isEmpty()) {
+                    return;
+                }
+
+                const pos = model.getPositionAt(edit.pos);
+                if (edit instanceof messages.Delete) {
+                    const endPos = model.getPositionAt(edit.pos + edit.length);
+                    monacoEdits.push({
+                        range: new monaco.Range(pos.lineNumber, pos.column, endPos.lineNumber, endPos.column),
+                        text: null
+                    });
+                } else if (edit instanceof messages.Insert) {
+                    monacoEdits.push({
+                        range: new monaco.Range(pos.lineNumber, pos.column, pos.lineNumber, pos.column),
+                        text: edit.content,
+                        forceMoveMarkers: true
+                    });
+                }
+            });
+
+            //this.editor.getModel().applyEdits(monacoEdits);
+            // TODO: above API call won't put the edits on the undo stack. Should other people's edits go on my undo stack?
+            //       below calls implement that. It is weird to have other peoples' edits in your undo stack, but it
+            //       also gets weird when they aren't – your undos get out of sync with the document.
+            //       Maybe there's something different that can be done with undo stops – I don't really know what they
+            //       are because it's not well documented (JS)
+            this.editor.pushUndoStop();
+            this.editor.executeEdits("Anonymous", monacoEdits);
```

Comment:
What is done with the `source` field here? Would we eventually want this to be populated with something like a user name or session id?

Reply:
We're also tracking a "name" on the server side, but it's also always set to "anonymous". I'm not sure if monaco can do something with it from a UI perspective, but I was just gonna wait to add the name to the message until there's a chance for it to be meaningful (i.e. some kind of auth or ID mechanism)